v2.0

kalaninja opened the pull request:
Hi there, 
I've managed to deal with all my "busy man" problems. So, back online and ready for some opensource.

Here is the PR for v2.0. It doesn't have a readme file yet, because I am unaware about the links to godoc, travis, etc. I assume that if we tag the new commit with as "v2.0", the link to godoc would be smth like 
https://godoc.org/gopkg.in/ahmetalpbalkan/go-linq.v2.0 and it seems that tha branch for travis and coveralls can be specified at their urls. Is it correct? Should I tag it myself?

ahmetalpbalkan commented:
@kalaninja thanks for this PR! Don't worry about README etc for now, we can address the docs/website issues later.

Do you mind giving a brief overview of how much exported function/type signatures have changed? That would help me to review and give an idea of how much of a breaking change it is.

ahmetalpbalkan commented on LICENSE:
```diff
@@ -1,14 +1,21 @@
-Copyright 2013 Ahmet Alp Balkan
+The MIT License (MIT)
```
I guess this one is a leftover. Can you please revert this file?

kalaninja replied:
Sure. Didn't notice that.

ahmetalpbalkan commented on example_test.go:
```diff
@@ -0,0 +1,226 @@
+	// [1 2 3 4 5 6 7 8 9 10]
+}
+
+type Book struct {
```
I believe this is unused

kalaninja replied:
It is an example for godoc

ahmetalpbalkan replied:
umm but it's not used in Example* functions either?

kalaninja replied:
This function is attached to godoc automatically. You can see the result in godoc for my current repo. https://godoc.org/github.com/kalaninja/go2linq#Query.ToSlice
godoc attaches this function to Query.ToSlice() and as the output is specified in the comment it is used as a test as well.

ahmetalpbalkan replied:
Yes but I still don't see the `Book` struct anywhere?

kalaninja replied:
Oh, I see! I probably wanted to make some example... Now this struct seems abandoned, I'll remove it.

ahmetalpbalkan commented on result.go:
```diff
@@ -0,0 +1,394 @@
+
+// ToChannel iterates over a collection and outputs each element
+// to a channel, then closes it.
+func (q Query) ToChannel(result chan<- interface{}) {
```
aren't you supposed to instantiate the chan with `new`?

kalaninja replied:
The idea is that my To*Smth*() methods do not create anything, but simply populate storages created by user instead. It was necessary for the end-user in order not deal with slice of interfaces, but to work with a typed collections. And I decided to use the same principle for the ToChannel() method to keep the API easier and more understandable.

ahmetalpbalkan replied:
ah sorry I reviewed this 1 AM yesterday, didn't realize this is actually an input parameter. I thought it's a named return.

kalaninja replied:
I understand you more that perfectly. Due to the time lag I usually answer you in the night)

kalaninja commented:
About an overview it seems to be a good idea? How do you want to have that info? I may do it as a table in excel or google docs with my methods on one side and yours on the other? Or you might have a better idea?

ahmetalpbalkan commented:
@kalaninja actually, I think I got the main idea. There are quite many new methods, some methods are gone, errors are gone, `linq.T` is now `interface{}` etc.

I also found godocs to be helpful. How about you add a `README.md` (by updating the old one?) in the meanwhile and I give you commit access after merging this. The remaining tasks would be:

- we should probably add a big notice to the README.md saying v2.0 is a breaking change, for those who want, we can link to a v1.0 branch
  - for that we first need to tag a v1.0 branch :) this can be done later.
- update README with new examples
- update Wiki tab with new examples (or maybe get rid of the wiki altogether, WDYT?)
- update the `gh-pages` branch for the website with new examples

kalaninja commented:
Yes, I don't have error because they made the predicates look ugly, and its effect can be achieved with panic and recover.
Yes, I used interface{} instead of linq.T, but now I am unsure about that decision, probably usage of linq.T made the code look prettier. WDYT?

I'll do the readme asap. Agree with the todo list. And I do not believe that the wiki is really helpful (I didn't even know that you have one)) You have a beautiful site and we might invest more effort there.

ahmetalpbalkan commented on compare.go:
```diff
@@ -0,0 +1,186 @@
+//		func (f foo) CompareTo(c Comparable) int {
+//			a, b := f.f1, c.(foo).f1
+//
+//			if a < b {
```
this code would be shorter as

```
switch{
case a>b: return -1
case b<a: return 1
default: return 0
}
```

kalaninja replied:
they seem pretty the same legth since I use gofmt
			switch {
			case a > b:
				return -1
			case b < a:
				return 1
			default:
				return 0
			}
vs
			if a < b {
				return -1
			} else if a > b {
				return 1
			}

			return 0

There are other variants as well. Consider:
			if a < b {
				return -1
			}
			if a > b {
				return 1
			}
			return 0
WDYT?

ahmetalpbalkan replied:
I guess it doesn't matter too much then. I still prefer

```
	switch {
	case a > b:
		return -1
	case b < a:
		return 1
	default:
		return 0
	}
```

mostly because it's easier to follow

kalaninja replied:
Yes, I agree. It looks really straightforward. I'll change that.

ahmetalpbalkan commented:
@kalaninja This generally looks good to me. Can you just add a README by copying from the current one and updating the example etc. Then it would be good to merge. Then we can address other stuff in https://github.com/ahmetalpbalkan/go-linq/pull/30#issuecomment-243571343.

kalaninja commented:
I've attached the readme, changed "if" to "switch" and removed the book struct from the example.

ahmetalpbalkan commented on README.md:
````diff
@@ -0,0 +1,63 @@
+	}).First()
+```
+
+**More examples** can be found in [documentation](https://godoc.org/gopkg.in/ahmetalpbalkan/go-linq.v2)
````
Maybe don't link to these like v2 yet because we'll merge it to `master`.

kalaninja replied:
And what will happen with the current version?

ahmetalpbalkan replied:
I archived the current version in `archive/0.9` branch, anybody who wants to use it I guess can grab it from there or perhaps grab it via `go get gopkg.in/...` (we have a `v0.9-rc5` tag, we can tag it as `v0.9` anytime if that'll make it work with gopkg.in).

ahmetalpbalkan commented on README.md:
````diff
@@ -0,0 +1,63 @@
+
+**Example:** Find an author that has written the most books
+```go
+import . "gopkg.in/ahmetalpbalkan/go-linq.v2"
````
We shouldn't suggest using gopkg.in.

kalaninja replied:
ok. I didn't know that

ahmetalpbalkan commented on README.md:
```diff
@@ -0,0 +1,63 @@
+* Supports arrays, slices, maps, strings, channels and
+custom collections (collection needs to implement Iterable interface
+and element - Comparable interface)
+* Parallel LINQ (PLINQ) *(comming soon)*
```
hmm maybe don't mention this just yet?

kalaninja replied:
Oh, sure. I also thought to delete this. Especially, because similar functionality can be achieved From(channel) and ToChannel() methods

ahmetalpbalkan commented on README.md:
```diff
@@ -0,0 +1,63 @@
+# go-linq [![GoDoc](https://godoc.org/gopkg.in/ahmetalpbalkan/go-linq.v2?status.svg)](https://godoc.org/gopkg.in/ahmetalpbalkan/go-linq.v2) [![Build Status](https://travis-ci.org/ahmetalpbalkan/go-linq.svg?branch=v2.0)](https://travis-ci.org/ahmetalpbalkan/go-linq) [![Coverage Status](https://coveralls.io/repos/github/ahmetalpbalkan/go-linq/badge.svg?branch=v2.0)](https://coveralls.io/github/ahme ...
+A powerful language integrated query (LINQ) library for Go.
```
I see you removed the notice about **written in vanilla Go**, doesn't this still hold true?

kalaninja replied:
Yes, that is still true, and sounds really cool. I'll bring it back

ahmetalpbalkan commented on README.md:
```diff
@@ -0,0 +1,63 @@
+# go-linq [![GoDoc](https://godoc.org/gopkg.in/ahmetalpbalkan/go-linq.v2?status.svg)](https://godoc.org/gopkg.in/ahmetalpbalkan/go-linq.v2) [![Build Status](https://travis-ci.org/ahmetalpbalkan/go-linq.svg?branch=v2.0)](https://travis-ci.org/ahmetalpbalkan/go-linq) [![Coverage Status](https://coveralls.io/repos/github/ahmetalpbalkan/go-linq/badge.svg?branch=v2.0)](https://coveralls.io/github/ahme ...
```
Hmm let's just remove all v2.0s here. It's not going to be a branch, I'm planning to make this live in master (and tag it as `v2.0` obviously) and archive the old version in a branch named `archive/0.9` if we need to take any fixes in the old code.

This code will be `master`, so we don't need to specify a branch name.

kalaninja replied:
Ok, I got that

ahmetalpbalkan commented on README.md:
```diff
@@ -0,0 +1,63 @@
+
+## Installation
+
+    $ go get gopkg.in/ahmetalpbalkan/go-linq.v2
```
Let's just use github here as well. I'm in the camp of people who are against pulling from gopkg.in as well. It's more likely that github will last longer :)

kalaninja replied:
Sounds reasonable. I'll fix everything connected with branching

ahmetalpbalkan commented on README.md:
```diff
@@ -0,0 +1,63 @@
+	authors []string
+}
+
+author := From(books).SelectMany(func(b interface{}) Query {
```
I find it a bit hard to follow this example 😵 . Not sure if it's just me but all the `a`s and `b`s kind of seemed confusing.

ahmetalpbalkan replied:
I've been thinking how to make this more understandable. Perhaps do not follow gofmt indentation rules and squeeze some funcs into single lines and do a custom indentation, and add some `// comments`?

kalaninja replied:
does it look better?

ahmetalpbalkan commented on README.md:
```diff
@@ -1,100 +1,63 @@
-
-[Ahmet Alp Balkan](http://ahmetalpbalkan.com) – [@ahmetalpbalkan](https://twitter.com/ahmetalpbalkan)
-
-## Release Notes
```
can you please  keep the Release Notes in the new README? and also add v2.0 briefly stating it is a breaking change and a total rewrite with greater performance, efficiency and new functionality?

kalaninja replied:
Do you want to keep all the previous notes?

kalaninja replied:
I'll put the release notes in the beginning, since for the the guys that already use go-linq  it is probably more important than the examples

ahmetalpbalkan replied:
Hmm you can leave it here. I was planning to add a big notice to the top about the breaking change after merging this.

kalaninja commented:
I've fixed links in readme, removed the line about plinq, recovered the one about vanilla go, added release notes.
I've tried several ways to make the example more readable and ended with renaming a to author, b to book and g to group. I've added comments as well.

ahmetalpbalkan commented:
@kalaninja thanks! I'll open issues for the rest of the updates we probably need to make.

ahmetalpbalkan commented:
💥 🎊 🍷
